sympy/functions/combinatorial/factorials.py

```diff
@@ -866,192 +959,505 @@ def fdiff(self, argindex=1):
+        from sympy import Product
+        n, k = self.args
+        # see if re-evaluation evaluates
+        e = self.func(n, k)
+        if e.func != self.func:
+            return e
+
+        # see if there is a non-negative k or n - k
+        if k.is_number and int_like(k) and k.is_nonnegative:
+            pass
+        elif (n - k).is_number and int_like(n - k) and (n - k).is_nonnegative:
             k = n - k
-
-        if k.is_Integer:
-            if k.is_zero:
-                return S.One
-            elif k.is_negative:
-                return S.Zero
-            else:
-                n, result = self.args[0], 1
-                for i in range(1, k + 1):
-                    result *= n - k + i
-                    result /= i
-                return result
         else:
-            return binomial(*self.args)
+            return self
+        i = Dummy('never_seen')
+        return (Product(n - k + i, (i, 1, k))/factorial(k)).doit()
+
+    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
+        from sympy import gamma, Abs, Lt
+        isint = lambda x: Eq(x, floor(x))
+        nneg = lambda x: Or(Eq(x, 0), Eq(x/Abs(x), 1))
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
+        return Piecewise(
+            (factorial(n)/factorial(k)/factorial(n - k),
+                And(isint(n), isint(k),
+                nneg(n), nneg(k), nneg(n - k))),
+            (ff(n, k)/factorial(k),
+                And(isint(k), nneg(k))),
+            (ff(n, n - k)/factorial(n - k),
+                And(isint(n - k), nneg(n - k))),
+            (0, And(*[i for x in (n, k, n - k)
+                for i in [isint(x), Lt(x, 0)]])),
+            (gamma(n + 1)/gamma(k + 1)/gamma(n - k + 1),
+                True))
 
     def _eval_rewrite_as_factorial(self, n, k, **kwargs):
-        return factorial(n)/(factorial(k)*factorial(n - k))
+        from sympy import Pow
+
+        if n.is_nonnegative and k.is_nonnegative or any(x is False for x in [n.is_integer, k.is_integer]):
+            if (n - k + 1).is_zero:
+                return S.Zero
+            return factorial(n)/(factorial(k)*factorial(n - k))
+        if fuzzy_and([n.is_integer, k.is_integer]):
+            if n.is_negative:
+                if k.is_nonnegative:
+                    return(Pow(-1, k)*factorial(k - n - 1)/(
+                        factorial(k)*factorial(-n - 1)))
+                if k.is_negative:
+                    return (Pow(-1, n - k)*factorial(-k - 1)/(
+                        factorial(n - k)*factorial(-n - 1)))
+            elif n.is_nonnegative:
+                if k.is_negative:
+                    return S.Zero
 
     def _eval_rewrite_as_gamma(self, n, k, **kwargs):
-        from sympy import gamma
-        return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
+        from sympy import gamma, Pow
+
+        if n.is_nonnegative and k.is_nonnegative or any(x is False for x in [n.is_integer, k.is_integer]):
```

Comment:
This is where we could handle the assumptions: if we change the `any` to `not all(i.is_integer for i in (n, k))` then lacking an assumption of integer would return gamma for consistent with nonnegative integers. Would 8016 pass if we made this change? This would mean that "binomial's expressed in terms of symbols not explicitly stated to be integers will be treated like nonnegative integers during rewriting."

I don't like this however. Another option is to recognize hints with the rewrites: 'combinatorial' or 'extended', perhaps?  Would a global keyword be better?

Reply:
ideally 8016 should only be valid for the Knuth implementation. I think @oscargus added that test in, we should ask him.
Personally I think adding support for 8016 is not entirely necessary.
[I tried in on Wolfram and it just gave a hypergeometric Function ](https://www.wolframalpha.com/input/?i=Simplify%28Sum%28binomial%28m%2C+k%29*binomial%28m%2C+n+-+k%29*%28-1%29**k%2C+%28k%2C+0%2C+n%29%29%2C+n+%3E+0+%29)

Reply:
>  This would mean that "binomial's expressed in terms of symbols not explicitly stated to be integers will be treated like nonnegative integers during rewriting.

This would go against our <b>rewrite only when the assumptions are valid</b> case too

Reply:
But it doesn't go against the policy of being as permissive as assumptions allow, e.g. `x/x` is 1 not `Piecewise((1, Ne(x, 0)), (Undefined, True))`.